Tools/ci-run.sh

```diff
@@ -3,125 +3,172 @@
   sudo apt-add-repository -y "ppa:ubuntu-toolchain-r/test"
   sudo apt update -y -q
   sudo apt install -y -q ccache gdb python-dbg python3-dbg gcc-$GCC_VERSION || exit 1
-  if [ -z "${BACKEND##*cpp*}" ]; then
+
+  ALTERNATIVE_ARGS=""
+  if [[ $BACKEND == *"cpp"* ]]; then
     sudo apt install -y -q g++-$GCC_VERSION || exit 1
+    ALTERNATIVE_ARGS="--slave /usr/bin/g++ g++ /usr/bin/g++-$GCC_VERSION"
   fi
   sudo /usr/sbin/update-ccache-symlinks
   echo "/usr/lib/ccache" >> $GITHUB_PATH # export ccache to path
 
-  sudo update-alternatives --install /usr/bin/gcc gcc /usr/bin/gcc-$GCC_VERSION 60 $(if [ -z "${BACKEND##*cpp*}" ]; then echo " --slave /usr/bin/g++ g++ /usr/bin/g++-$GCC_VERSION"; fi)
+  sudo update-alternatives --install /usr/bin/gcc gcc /usr/bin/gcc-$GCC_VERSION 60 $ALTERNATIVE_ARGS
 
   export CC="gcc"
-  if [ -z "${BACKEND##*cpp*}" ]; then
+  if [[ $BACKEND == *"cpp"* ]]; then
     sudo update-alternatives --set g++ /usr/bin/g++-$GCC_VERSION
     export CXX="g++"
   fi
-fi
-if [ "${OS_NAME##macos*}" == "" ]; then
+elif [[ $OSTYPE == "darwin"* ]]; then
+  echo "Setting up macos compiler"
   export CC="clang -Wno-deprecated-declarations"
   export CXX="clang++ -stdlib=libc++ -Wno-deprecated-declarations"
 fi
 
 # Set up miniconda
-if [ "$STACKLESS" == "true" ]; then
+if [[ $STACKLESS == "true" ]]; then
   echo "Installing stackless python"
   #conda install --quiet --yes nomkl --file=test-requirements.txt --file=test-requirements-cpython.txt
   conda config --add channels stackless
   conda install --quiet --yes stackless || exit 1
 fi
 
+PYTHON_SYS_VERSION=$(python -c 'import sys; print(sys.version)')
+
 # Log versions in use
 echo "===================="
 echo "|VERSIONS INSTALLED|"
 echo "===================="
-python -c 'import sys; print("Python %s" % (sys.version,))'
-if [ "$CC" ]; then
+echo "Python $PYTHON_SYS_VERSION"
+if [[ $CC ]]; then
   which ${CC%% *}
   ${CC%% *} --version
 fi
-if [ "$CXX" ]; then
+if [[ $CXX ]]; then
   which ${CXX%% *}
   ${CXX%% *} --version
 fi
 echo "===================="
 
 # Install python requirements
 echo "Installing requirements [python]"
-if [ -z "${PYTHON_VERSION##2.7}" ]; then
+if [[ $PYTHON_VERSION == "2.7"* ]]; then
   pip install wheel || exit 1
   pip install -r test-requirements-27.txt || exit 1
-elif [ -z "${PYTHON_VERSION##3.[45]*}" ]; then
+elif [[ $PYTHON_VERSION == "3."[45]* ]]; then
   python -m pip install wheel || exit 1
   python -m pip install -r test-requirements-34.txt || exit 1
 else
   python -m pip install -U pip setuptools wheel || exit 1
 
-  if [ -n "${PYTHON_VERSION##*-dev}" ]; then
+  if [[ $PYTHON_VERSION == *"-dev" || $COVERAGE == "1" ]]; then
     python -m pip install -r test-requirements.txt || exit 1
 
-    if [ "${PYTHON_VERSION##pypy*}" -a "${PYTHON_VERSION##3.[4789]*}" ]; then
+    if [[ $PYTHON_VERSION == "pypy"* && $PYTHON_VERSION == "3."[4789]* ]]; then
       python -m pip install -r test-requirements-cpython.txt || exit 1
     fi
```

Comment:
The two conditions are still the wrong way around here. (Looks like my [comment](https://github.com/cython/cython/pull/4324#discussion_r684609533) was lost due to later changes.)

Reply:
Yes, I translated this incorrectly to the new syntax. But then it will be triggered only for cpython 3.10-dev (all non `-dev` versions are cut off by earlier `$PYTHON_VERSION == *"-dev"` and we add `COVERAGE = 1` only on cpython  3.7 and 3.9) is it the correct behavior?

Reply:
Right, the `-dev` test is also reversed. Note the difference between `-n` (does not match the replacement regex) and `-z` (matches the regex).

Reply:
> Note the difference between `-n` (does not match the replacement regex) and `-z` (matches the regex).

Yes, just at that time it was new to me and I was confused about it.

Reply:
[resolved](https://github.com/cython/cython/pull/4324/commits/2ae2103173afa4bc7416eded71adb66c642e9f06)